### PR TITLE
refactor(cli): split the 2 big functions into small ones

### PR DESCRIPTION
**Pull request by cmalard**
+ use `path.resolve()`
+ remove apparently useless? guard for Mocha

### CODE REVIEW

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: We want to keep these auxiliary functions as variable declarations.

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: How positive is it that there are no side effects by redefining this even in mocha runs? The original code seemed to guard this being added to only when it was needed. I'm wary of removing that guard without knowing it is safe.

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: This return while it looking lean also looks like it could be easy to accidentally break later. I'd rather we had dedicated if statements to return in the different contexts.

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: I'd prefer it if we started using `path.resolve()` instead of manually making directory/file paths. This way the separators are OS defined (using `\` for Windows for example) and less likely to cause quirkyness for users in case of problems.

**Review by Garbee — CHANGES_REQUESTED** [review]
Along with the inline comments. Please drop adding the node package lock. That has nothing to do with the CLI refactoring. Adding in auxiliary tooling files should be done in their own PRs.

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: This is to be consistent with every other files of this project, why would we want to keep those different ?

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: This converts to:
```js

  var commit = getContentFromFile(msgFileOrText);
  if (commit) {
    return commit;
  }

  if (!msgFileOrText) {
    commit = getContentFromFile('COMMIT_EDITMSG');
    if (commit) {
      return commit;
    }
  }

  return { message: msgFileOrText };
```
Do you see any way to shorten it ?

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: do you want this to have its own commit ?

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: I tried to find what could happen.

The guard was:
```js
if (process.argv.join('').indexOf('mocha') === -1) {}
```

So I added at the beginning of the file:
```js
console.warn('process.argv:', process.argv)
```
and started `npm run test` - result:
```
      process.argv: [ '/usr/bin/nodejs',
        '/home/cedric/Projects/validate-commit-msg/lib/cli.js' ]
```
Can't reproduce this case, looks like it was always going inside this `if`.
Do you have some clues to reproduce ? @kentcdodds ?

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: This should be functionally equivalent. But I'd also not use the logical not operator. Instead hard do the checks to `null` as that is returned if not found. Minuscule optimization to save the engine some time. But also providing more legible code in the future since you know exactly what the check is looking for all the time. 

```javascript
var commit = getContentFromFile(msgFileOrText);

if (commit === null) {
  commit = getContentFromFile('COMMIT_EDITMSG');
}

if (commit === null) {
  commit = { message: msgFileOrText };
}

return commit;
```

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: Since you're touching the code here refactoring anyways, just carry it out with this PR. It counts towards the refactor. :wink:

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: Those will eventually be converted as well. This way when ES6 is moved up to they can become constants to help ensure the functions don't get redefined.

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: Looking into the history Kent is the one who originally wrote the guard in. He's caught up in a family emergency so I wouldn't wait on a response.

From the looks of it the guard was in place since `index.js` also used to be the CLI endpoint. When that was moved out, the guard was kept possibly unknowingly that the move resolved the need for the guard.

So let's go ahead and remove it. If something breaks we know how to address it quickly.

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: As long as we're not using ES6, we should choose consistency.

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: Why removing `!msgFileOrText` on the second `if` ? We need it :-)
Ok with this structure :+1:

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: Then the rest of the system should get switched to functions being declared as vars.

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: That's not what the `msgFileOrText` was used to call for according to your original code. In your original code if it doesn't exist it then retrieved the `COMMIT_EDITMSG` data. Since that is setting the commit to `null` if not found, we only need to check for null against commit again. `msgFileOrText` is not useful in any checks logically.

Comment (Garbee) [inline lib/cli.js]: If you want to keep the check on the parameter this is how you'd do it as I'm understanding the original short code:

```javascript
if (msgFileOrText === undefined || msgFileOrText === null) {
	return getContentFromFile('COMMIT_EDITMSG');
}

var commit = getContentFromFile(msgFileOrText);

if (commit === null) {
	commit = { message: msgFileOrText };
}

return commit;
```

Also, this is exactly why I want the longer version. Easier to understand then deciphering the short clever version. Don't be clever. Clever is harder to maintain.

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: I'm not confortable with this style based [on this](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#function-declarations-to-hide-implementation-details-1)
Could you tell me what's the advantages ?

Comment (Garbee) [inline lib/cli.js]: I'm not sure why you're referencing the Angular styleguide. Is that mentioned anywhere in the project currently that it is what is followed? This is style, it all depends on what the teams needs are, there tooling, and how they want things to look. There is no right or wrong except considering the circumstances.

Comment (cmalard) [inline lib/cli.js]: The link is not here for Angular but for the arguments. Ok for your point on the style debate.
The fact is, actually most of the project is written with functions declarations. I like the advantages that comes with.
With this PR I choose consistency.
So you ask me to change all the files to functions expressions and I ask you what's the advantages ? :-)

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: wow, no at all. You made something really different there.
What if `COMMIT_EDITMSG` doesn't exist ?

I've rewrite it this way :
```js
var commit = getContentFromFile(msgFileOrText);
if (commit === null && !msgFileOrText) {
  commit = getContentFromFile('COMMIT_EDITMSG');
}
if (commit === null) {
  commit = { message: msgFileOrText };
}
return commit;
```
for comparison :
```js
  return getContentFromFile(msgFileOrText)
    || !msgFileOrText && getContentFromFile('COMMIT_EDITMSG')
    || { message: msgFileOrText };
```

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: The advantage is what I'm seeing as the future style of the project is going to be the `var/let/const` declarations on these things. So, starting things now to be towards that is better. Without a linter in place to enforce this, the consistency isn't *too* massive of an issue for me right now.

**Comment by cmalard** [comment]
@Garbee all your requests are adressed

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: Why check the parameter again here? And even if it is required, don't use the logical not operator.

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: Once again, let's not use logical operators on things.

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: Logical not operator again.

File: lib/cli.js
Comment (Garbee) [inline lib/cli.js]: Logical not operator not desired.

**Review by Garbee — CHANGES_REQUESTED** [review]

File: lib/cli.js
Comment (cmalard) [inline lib/cli.js]: again ? this is the only place.
cf https://github.com/conventional-changelog/validate-commit-msg/pull/109#discussion_r130929896

This is really annoying to have all this change requests about style.
The logical not operator is used in other places in the project.
Until there is ESLint rules in place, you should stop doing that...

Comment (Garbee) [inline lib/cli.js]: Hmm, didn't see that comment since the change was made and pushed before I checked the PR again.

So there was a bug in the code I didn't see. Hence why the cleverness of your original design was flawed for maintainability.

Instead of bickering over why you feel things are inconsistent and attempting to use my own words back at me, accept that proper tooling isn't in place for consistency checks across the whole system. The previous maintainer had his way of wanting code styled and didn't have the tooling setup to enforce it. Now, I'm the lead maintainer and I'd like us to move to a more declarative structure. It's that simple and nothing to be upset over.

Comment (cmalard) [inline lib/cli.js]: Get it, but please consider that as long as proper tooling is in place, you should stop nitpicking. This is really discouraging contributions.
This PR is not for style but for refactor.

Comment (Garbee) [inline lib/cli.js]: Ok, that's a valid point. So ignore the operators.

 But I'm still not understanding why you're checking for the parameter's existence at all. If you've already passed it into the function call and it returns null, it isn't relevant for the next possible commit value.

Comment (cmalard) [inline lib/cli.js]: Because at this point we have no idea if the parameter is a file or a text. Hence its name.
So we start checking if it's a valid file.
If not, it is a text.
In case this parameter is not provided, we fall back to COMMIT_EDITMSG.

Comment (Garbee) [inline lib/cli.js]: This is why declarative while being a little bit more verbose is super useful to maintain software. It makes you really think about the full flow required. :wink:

I see the bug in my port, which is easy enough to fix. I'm just not sure why you're thinking the check on the parameter internally is ever needed. Once it is used without a check, checking it beyond that isn't relevant given the code flow as I'm understanding it.

Comment (cmalard) [inline lib/cli.js]: Ok, took a step back to rethink about it.
- we actually have 2 cases : `msgFileOrText` is provided as an argument to the cli or not
- if not, we want to look at `COMMIT_EDITMSG`
- the function must return an object containing `message` key, whatever the case

To answer your question: the check is needed because if `msgFileOrText` is provided, we do not want to default on `COMMIT_EDITMSG`.

I refactored it to make all this clearer:
```js
  if (msgFileOrText !== undefined) {
    return getCommitFromFile(msgFileOrText) || { message: msgFileOrText };
  }

  return getCommitFromFile('COMMIT_EDITMSG') || { message: null };
```
Now for the pleasure we can shorten it a bit: it's not important here if `message` is `null` or `undefined`, we did it both in our codes. So the fallbacks can be merged:
```js
  return getCommitFromFile(msgFileOrText || 'COMMIT_EDITMSG')
    || { message: msgFileOrText };
```

All these codes have the exact same result, even [those](https://github.com/conventional-changelog/validate-commit-msg/pull/109#discussion_r130929896).

So let's keep the first one in this message, it's maybe the simplest to understand :beers:
And if we can auto-check `GITGUI_EDITMSG`, it will be the easiest to extend :wink:

Comment (Garbee) [inline lib/cli.js]: Yup, the first one looks perfectly reasonable. I'd simply add an extra not `null` check just in case to the statement. Other than that nothing too tricky going on to keep things super-concise. Easy to tell what's expected.

Very good work figuring that one out. You explained the logic needs very well.

Comment (cmalard) [inline lib/cli.js]: Hi @Garbee 
`null` is not desired here, the case `undefined` equals "no message provided".
If we get `null`, it means it was desired.
Looks good to you ? Before going on deprecation we could merge this PR.

**Pull request merged**
